presto-hive/src/main/java/io/prestosql/plugin/hive/SyncPartitionMetadataProcedure.java

```diff
@@ -0,0 +1,224 @@
+                    .collect(toImmutableList());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean validatePath(FileSystem fileSystem, FileStatus file, Column column)
+    {
+        try {
+            Path path = file.getPath();
+            String prefix = column.getName() + '=';
+            return fileSystem.isDirectory(path) && path.getName().startsWith(prefix);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    // calculate relative complement of set b with respect to set a
+    private Set<String> difference(List<String> a, List<String> b)
+    {
+        return Sets.difference(newHashSet(a), newHashSet(b));
+    }
+
+    private void addPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Path location,
+            Set<String> partitions)
+    {
+        partitions.forEach(name -> metastore.addPartition(
+                session,
+                table.getDatabaseName(),
+                table.getTableName(),
+                buildPartitionObject(session, table, name, new Path(location, name)),
+                new Path(location, name),
+                PartitionStatistics.empty()));
+    }
+
+    private void dropPartitions(
+            SemiTransactionalHiveMetastore metastore,
+            ConnectorSession session,
+            Table table,
+            Set<String> partitions)
+    {
+        partitions.forEach(name -> metastore.dropPartition(
+                session,
+                table.getDatabaseName(),
+                table.getTableName(),
+                extractPartitionValues(name)));
+    }
+
+    private Partition buildPartitionObject(ConnectorSession session, Table table, String partitionName, Path targetPath)
+    {
+        return Partition.builder()
+                .setDatabaseName(table.getDatabaseName())
+                .setTableName(table.getTableName())
+                .setColumns(table.getDataColumns())
+                .setValues(extractPartitionValues(partitionName))
+                .setParameters(ImmutableMap.of(PRESTO_QUERY_ID_NAME, session.getQueryId()))
+                .withStorage(storage -> storage
+                        .setStorageFormat(isRespectTableFormat(session) ?
+                                table.getStorage().getStorageFormat() :
+                                fromHiveStorageFormat(HiveSessionProperties.getHiveStorageFormat(session)))
+                        .setLocation(targetPath.toString())
+                        .setBucketProperty(table.getStorage().getBucketProperty())
+                        .setSerdeParameters(table.getStorage().getSerdeParameters()))
+                .build();
+    }
+
+    SyncMode toSyncMode(String mode)
+    {
+        try {
+            return SyncMode.valueOf(mode);
```

Comment:
We should call `mode.toUpperCase(ENGLISH)` here, as we do for other enums like table format.